services/QuillLMS/client/app/bundles/Shared/components/renderForQuestions/feedback.tsx

```diff
@@ -114,14 +132,119 @@ function getCSSClasses(feedbackType: string): string {
-    <div className='feedback-row student-feedback-inner-container'>
-      <img alt={getIconAlt(feedbackType)} className={getIconClassName(feedbackType)} src={getFeedbackIcon(feedbackType)}  />
-      {feedback}
+interface FeedbackProps {
+  correctResponse?: string,
+  feedbackType: any,
+  feedback: any,
+  latestAttempt?: any,
+  question?: any,
+  showTranslation?: boolean,
+  translate?: (input: string) => string
+}
+
+function translatedCMSResponseFeedback(latestAttempt, question) {
+  const { response } = latestAttempt
+  const { translation } = question
+  const { cms_responses } = translation
+  const id = response.id || response.parent_id
+  const translatedResponse = cms_responses[`cms_responses.${id}`]
+  return translatedResponse ? ReactHtmlParser(translatedResponse) : null
+}
+
+function translatedIncorrectContinueFeedback(translate, latestAttempt, correctResponse) {
+  const firstPhrase = translate('feedback^Good try!')
+  const secondPhrase = translate('feedback^Compare your response to the strong response, and then go on to the next question')
+  const thirdPhrase = translate('feedback^Your response:')
+  const fourthPhrase = translate('feedback^A strong response:')
+  return `<b>${firstPhrase}</b>${secondPhrase}<br><br><b>${thirdPhrase}</b><br>${latestAttempt}<br><br><b>${fourthPhrase}</b><br>${correctResponse}`
+}
+
+function translatedReviseMatchedFeedback(question, latestAttempt, translate) {
+  const { response } = latestAttempt
+  if (response?.isIncorrectSequence && question?.translation?.incorrectSequences) {
+    const { uid } = response
+    const { translation } = question
+    const { incorrectSequences } = translation
+    const key = `incorrectSequences.${uid}`
+    return incorrectSequences[key] ? ReactHtmlParser(incorrectSequences[key]) : null
+  }
+  if (response?.isFocusPoint && question?.translation?.focusPoints) {
+    const { uid } = response
+    const { translation } = question
+    const { focusPoints } = translation
+    const key = `focusPoints.${uid}`
+    return focusPoints[key] ? ReactHtmlParser(focusPoints[key]) : null
+  }
+  if(response?.author) {
+    const { author, feedback } = response
+    if(author === "Quotation Mark Hint") {
+      return(
+        <span>
+          <br />
+          <p>{translate('feedback^It looks like you might have used two apostrophes to make a quotation mark.')}</p>
+          <p>{translate('feedback^Instead of hitting the apostrophe key twice to make a quotation mark, hold down the shift key and hit the apostrophe key once.')}</p>
+        </span>
+      )
+    }
+    if(C.ERROR_AUTHORS.includes(author)) {
+      return translate(`feedback^${feedback}`)
+    }
+  }
+}
+
+const Feedback = ({
+  correctResponse,
+  feedbackType,
+  feedback,
+  latestAttempt,
+  question,
+  showTranslation,
+  translate
+}: FeedbackProps) => {
+
+  function getTranslatedFeedback(showTranslation, question, feedbackType) {
```

Comment:
I'm not sure exactly how to clean this up, but this method feels difficult to parse to me, and it's not clear where the logic for the ordering comes from (ie, if I had a new feedback type to add or something, I wouldn't know where to put it). I think it might help just to do a series of early returns rather than the long if/else statement. I'd add this function:

```
function renderFeedback(feedback) {
    if(feedback && typeof feedback === 'string') {
      feedback = <p>{feedback}</p>
    }
    return feedback
}
```

And then do something like this inside of `getTranslatedFeedback`:

```
if (question?.translation?.instructions && showInstructions) {
      const { translation } = question
      const { instructions } = translation
      return renderFeedback(feedback)
}

if (feedbackType === DEFAULT_FILL_IN_BLANK) {
      return renderFeedback(translate('feedback^Fill in the blank with the correct option.'))
}
....
```

And then maybe add some comments if it seems helpful?

Reply:
ok, I'll tweak this function a bit!